### PR TITLE
Migrate to Null Safety

### PR DESCRIPTION
**Pull request by caseycrogers**
In progress PR for migrating to Null Safety.

@synw I had an API design question: should users be able to create data frames with:
* some columns unspecified in some rows?
* some column values explicitly set to null?
Eg is the following the intended behavior:
```
    final rows = <Map<String, dynamic>>[
      <String, dynamic>{'col1': 'a', 'col2': 1, 'col3': 1.0, 'col4': date},
      <String, dynamic>{'col1': 'b', 'col2': 2, 'col3': 2.0, 'col4': date},
      // col3 is null and col4 is unspecified
      <String, dynamic>{'col1': 'c', 'col2': 3, 'col3': null},
    ];
    df = DataFrame.fromRows(rows);
    expect(df.dataset, <dynamic>[
      <dynamic>['a', 1, 1.0, date],
      <dynamic>['b', 2, 2.0, date],
      // null value is passed through, unspecified value is converted to null
      <dynamic>['c', 3, null, null],
    ]);
```

### CODE REVIEW

**Comment by caseycrogers** [comment]
@synw  (theoretically) all done! I'll spend a day or two spiffing it up and adding some unit tests to further verify proper null handling. Separate idea, let me know your thoughts:
Maybe `df` should have a constructor argument for nullable vs non-nullable. If a user doesn't want to allow null values, we should throw any time they add a row with an omitted column or add a value with a null value.

**Comment by caseycrogers** [comment]
Oof this has become a huge refactor. All tests currently passing but the sort test. It's not passing because the old behavior (and the test) assumed greatest-to-least sort order. I've updated it to use the default dart sort implementation which assumes least-to-greatest sort order. Do we want to:
a. Update the test to expect least-to-greatest sort order
b. Update the implementation to maintain greatest-to-least sort order
or
c. Add a sort order argument where the user can specify their desired sort order

File: lib/src/df.dart [hunk 1]
```diff
@@ -31,12 +29,15 @@ class DataFrame {
   // ********* data **********
 
   /// An iterable of rows data
-  Iterable<Map<String, dynamic>> get rows => _iterRows();
```
Comment (caseycrogers) [inline lib/src/df.dart:34]: I've replaced `dynamic` with `Object` wherever it appears. Dynamic allows you to call any method without a compile time error. If the runtime type doesn't have that method, you'll get a runtime error. Object does not allow you to call arbitrary methods and forces you to explicitly cast the variable first (which can cause a runtime error, but at least it's forcing you to be explicit when you're doing something dangerous). Additionally, `dynamic` is implicitly nullable whereas `Object` is non-nullable by default and must be declared nullable.

Because of the above two reasons, code using `Object` is safer and easier to read so I've switched to it.

File: lib/src/df.dart [hunk 2]
```diff
@@ -51,33 +52,34 @@ class DataFrame {
       List<String>.from(_columns.map<String>((c) => c.name));
 
   /// The dataframe columns indices
-  Map<int, String> get columnsIndices => _columnsIndices();
```
Comment (caseycrogers) [inline lib/src/df.dart:54]: I don't think this data structure needed to be a map-I've changed it to a list to improve readability.

File: lib/src/df.dart [hunk 3]
```diff
@@ -158,7 +160,7 @@ class DataFrame {
             timestampFormat: timestampFormat);
         df._matrix.data.add(colValues);
       }
-      ++i;
```
Comment (caseycrogers) [inline lib/src/df.dart:161]: `i++` is more idiomatic when you don't need to assign the pre-increment value to a variable so I've changed to it.

File: lib/src/df.dart [hunk 4]
```diff
@@ -355,82 +351,110 @@ class DataFrame {
...
-    for (final col in _columns) {
-      ind[i] = col.name;
-      ++i;
-    }
-    return ind;
+  List<String> _columnIndices() => columns.map((c) => c.name).toList();
+}
+
+/// How to treat nulls when taking the average over a column.
+enum NullMeanBehavior {
```
Comment (caseycrogers) [inline lib/src/df.dart:456]: I've added arguments to `sort` and `mean` to require users to explicitly specify how nulls should be handled when aggregated a column.

File: lib/src/matrix.dart [hunk 5]
```diff
@@ -1,86 +1,83 @@
...
+    final dataRows = <Map<String, Object>>[];
     for (final row in data.sublist(startIndex, endIndex)) {
-      final dataRow = <String, dynamic>{};
+      final dataRow = <String, Object>{};
       var i = 0;
-      row.forEach((dynamic item) {
-        dataRow[indices[i]] = item;
-        ++i;
+      row.forEach((Object? item) {
+        if (item != null) dataRow[columnNames[i]] = item;
```
Comment (caseycrogers) [inline lib/src/matrix.dart:39]: Rather than returning a map with nulls, we simply return a map with null entries unspecified. This means we have no way to distinguish between a value that was never specified or a value that was explicitly specified as null. If we want to change this this will require us adding an extra sentinel type for "not specified" or something-this would be a significant refactor. Let me know if this is desirable.

File: lib/src/df.dart (hunk 4 above)
Comment (synw) [reply lib/src/df.dart:456]: I guess we will have to document these changes when released

**Comment by synw** [comment]
> Maybe df should have a constructor argument for nullable vs non-nullable

Good idea, so that the user can control this behavior. I am pretty much ignorant about null safety migration as I don't do much Dart these days and have never migrated any libs to null safety yet. You are doing an excellent work on this lib, thanks to you. Do you want to be maintainer?

**Comment by caseycrogers** [comment]
Being a maintainer would be great, thanks!

Once this PR's in I'll make a follow up to enable users to construct a non-nullable DataFrame.

**Comment by synw** [comment]
@caseycrogers I sent you an invitation to join the repository: I think you must approve it so I can grant you the maintainer role

**Comment by oleggreen** [comment]
Hi guys, can you say when we might get a null safety for this plugin?

**Comment by caseycrogers** [comment]
@marchelo Hey! Sorry for the late response. I'll aim to have a prerelease out by Friday!

It's already 80% done I just need to update sorting.

**Comment by caseycrogers** [comment]
@synw this PR is done and I'm going to merge it. I've set the version to `0.4.0-nullsafety.0` to indicate that this is a prerelease-though I don't expect any major bugs. In addition to the unit tests passing, I've updated my personal project that depends on DF to use it and haven't had any issues.

Once it's merged could you update the pub.dev package? Or if you don't mind adding me as an uploarder to the pub.dev package, I'm happy to update it myself.

**Comment by caseycrogers** [comment]
fyi @marchelo this PR is done. It's not launched yet on pub.dev, but if you'd like to migrate to null safety sooner you can fork this repo and depend on it via a path dependency. Assuming you download `df` to the same parent directory as your project, update your pubspec.yaml to include the following:
```
  df:
    path: '../df'
```

Once this version is released to pub.dev you can go back to depending on it.

**Comment by oleggreen** [comment]
Thank You! @caseycrogers
